linked_list/implementation.py

```diff
@@ -1,40 +1,149 @@
+            self.start = self.end = None
+        else:
+            self.start = new_node = self.end = Node(elements[0])
+            self.length = 1     # for creating self.start
+            
+            for i in elements[1:]:
+                self.length += 1
+                new_node = Node(elements[i-1])
+                self.end.next = new_node        # new node is the new tail
+                self.end = self.end.next        # moving tail to the end
+                self.end.next = None
+                
+    def __iter__(self):
+        # should return something iterable
+        it = self.start       # to set the iterator to start, beginning
+        
+        while it:
+            yield (it.elem)
+            it = it.next
+        
+        raise StopIteration
+    
+        
     def __str__(self):
-        pass
+        result_list_string = None
+        if self.start == None:
+            result_list_string = "[]"
+        else:
+            result_list_string = "["
+            self.iterator = self.start
+            
+            # instead of this use iterators
+            while self.iterator is not None:
+                result_list_string += str(self.iterator.elem) + ", "
+                self.iterator = self.iterator.next
+            
+            result_list_string = result_list_string[:-2] + "]"
+        return result_list_string
 
     def __len__(self):
-        pass
-
-    def __iter__(self):
-        pass
+        return self.length
 
     def __getitem__(self, index):
-        pass
-
+        if self.start == None or index >= self.length:
+            raise IndexError
+        return list(iter(self))[index]
+        
     def __add__(self, other):
-        pass
-
+        # for adding two linked lists, new_list = my_list + LinkedList([1])
+        if self.start == None:
+            self.end = other.start
+        else:
+            self.end.next = other.start
+        
     def __iadd__(self, other):
-        pass
+        # operator.__iadd__(a, b)
+        # a = iadd(a, b) is equivalent to a += b
+        self.__add__(other)
 
     def __eq__(self, other):
-        pass
+        return str(self) == str(other)
+        
+    def __ne__(self, other):
+        # the opposite of __eq__()
+        return not self.__eq__(other)
 
     def append(self, elem):
-        pass
+        # adding new node at the end
+        new_node = Node(elem)
+        new_node.next = None
+        
+        if self.start == None:
+            self.start = new_node
+        else:
+            self.start.next = self.end = new_node
+        self.length += 1
+    
+
 
     def count(self):
-        pass
+        return self.length
 
     def pop(self, index=None):
-        pass
+        if self.length == 0 or index < 0 or index >= self.length:
+            raise IndexError
+        
+        self.length -= 1
+        
+        if index == None:
+            value = self[-1]
+            #remove last node here
```

Comment:
Removing the last node basically means looking at the second to last node and setting its tail to None (and making sure that self.end points to the new last node)
Well, plus some special cases for stuff like if there's only one node.

Make sense?